src/Icicle/Core/Exp/Simp.hs

```diff
@@ -0,0 +1,139 @@
+import qualified Icicle.Common.Exp.Simp.Beta    as B
+import           Icicle.Common.Fresh
+import           Icicle.Common.Type
+import qualified Icicle.Core.Exp                as C
+import           Icicle.Core.Exp.Prim
+
+import           P
+
+
+-- | Core Simplifier:
+--   * a normal
+--   * beta reduction
+--   * constant folding for some primitives
+--   * ...something exciting???
+--
+simp :: Ord n => (C.Exp n -> Bool) -> C.Exp n -> Fresh n (C.Exp n)
+simp isValue = anormal . simpX isValue
+
+
+simpX :: Ord n => (C.Exp n -> Bool) -> C.Exp n -> C.Exp n
+simpX isValue = go
+  where
+    beta  = B.beta isValue
+    go xx = case beta xx of
+      -- * constant folding for some primitives
+      XApp{}
+        | Just (p, as) <- takePrimApps xx
+        , Just args    <- sequenceA (fmap takeValue as)
+        -> fromMaybe xx (simpP p args)
+
+      -- * beta reduce primitive arguments
+      XApp{}
+        | Just (p, as) <- takePrimApps xx
+        -> makeApps (XPrim p) (fmap go as)
+
+      XApp p q
+        -> XApp (go p) (go q)
+
+      XLam n t x1
+        -> XLam n t (go x1)
+
+      XLet n x1 x2
+        -> XLet n (go x1) (go x2)
+
+      b@(XVar{})   -> b
+      b@(XPrim{})  -> b
+      b@(XValue{}) -> b
+
+
+-- | Primitive Simplifier
+--
+simpP :: Prim -> [(ValType, BaseValue)] -> Maybe (C.Exp n)
+simpP = go
+  where
+    go pp args = case pp of
+      PrimMinimal mp  -> simpMP mp args
+      -- * leaves fold and map alone for now..?
+      PrimFold    _ _ -> Nothing
+      PrimMap     _   -> Nothing
+
+simpMP :: M.Prim -> [(ValType, BaseValue)] -> Maybe (C.Exp n)
+simpMP = go
+  where
+    go pp args = case pp of
+      -- * arithmetic on constant integers
+      M.PrimArith M.PrimArithPlus
+        -> arith2 pp args (+)
+      M.PrimArith M.PrimArithMinus
+        -> arith2 pp args (-)
+      M.PrimArith M.PrimArithDiv
+        -> arith2 pp args div
+      M.PrimArith M.PrimArithMul
+        -> arith2 pp args (*)
+      M.PrimArith M.PrimArithNegate
+        | [(_, VInt x)] <- args
+        -> let t = functionReturns (M.typeOfPrim pp)
+           in  return $ XValue t (VInt (-x))
+
+      -- * predicates on integers
+      M.PrimRelation M.PrimRelationGt IntT
+        -> int2 args (>)
+      M.PrimRelation M.PrimRelationGe IntT
+        -> int2 args (>=)
+      M.PrimRelation M.PrimRelationLt IntT
+        -> int2 args (<)
+      M.PrimRelation M.PrimRelationLe IntT
+        -> int2 args (<)
+      M.PrimRelation M.PrimRelationEq IntT
+        -> int2 args (==)
+      M.PrimRelation M.PrimRelationNe IntT
+        -> int2 args (/=)
+
+      -- * logical
+      M.PrimLogical M.PrimLogicalAnd
+        -> bool2 args (&&)
+      M.PrimLogical M.PrimLogicalOr
+        -> bool2 args (||)
+      M.PrimLogical M.PrimLogicalNot
+        -> bool1 args not
+
+      -- * leaves baked-in constants and datetime alone
+      M.PrimConst    _ -> Nothing
+      M.PrimDateTime _ -> Nothing
+
+    bool1 args f
+      | [(_, VBool x)] <- args
```

Comment:
partial pattern match?! this is very naughty.
we should probably return an eval error in this case, but I think at the very least returning the original expression would be better because later typechecking or real eval would give a meaningful error